File: lc/reverse-bits.go
// n   0 1 0 0 0 1 1

// ans 1 1 0 0 0 0 0

// do this 31 times
// v = n & 1 // v = 1 or 0
// ans =>> 1
// ans |= mask // mask is 2**31. applied only if v != 0
// n =>> 1

// n    0 0 0
// ans  1 1 0 
// v    0 0 1
// mask 1 0 0

// 0 0 0 0
// n 0 0 0 0
// a 0 1 0 0
// m 0 0 1 0
// 31 --> 3
// i = 0, 1,  2 
func reverseBits(n int) int {
    ans, mone := 0, 1 << 30
    n >>= 1
    for _ = range 30 {
        if n & 1 != 0 {
            ans |= mone
        }
        mone >>= 1
        n >>= 1
    }
    return ans
}

File: lc/py/reverse-bits.py
# 011010

# 011 -- n
# 000 -- np
# 110 -- ans

#  ans

# 01

# 01

# ans = 0
# repeat 2 times
#   get last digit from n
#   apply it to ans -- add 1 or not depending on if n is even or odd
#   shift ans to left
#   shift n to right

# n 011
# n 000
# a 110
class Solution:
    def reverseBits(self, n: int) -> int:
        ans = 0
        for _ in range(31):
            if n % 2 == 1:
                ans += 1
            ans <<= 1
            n >>= 1
        return ans